File: playwright-java/src/test/java/testtracks/data/builders/restfulbooker/RoomDataBuilder.java
package testtracks.data.builders.restfulbooker;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import testtracks.data.JsonPayloads;
import testtracks.data.TestData;

public final class RoomDataBuilder {
  private static final ObjectMapper JSON = new ObjectMapper();
  private final ObjectNode payload;

  public RoomDataBuilder(RoomDataDefaults defaults) {
    payload = defaults.payload().deepCopy();
  }

  public static RoomDataBuilder forScenario(String scenarioId, RoomDataDefaults defaults) {
    var roomNumber =
        TestData.numericSuffix(defaults.roomIdMinimum(), defaults.roomIdMaximumExclusive());
    var builder = new RoomDataBuilder(defaults);
    builder
        .withRoomId(roomNumber)
        .with("description", builder.stringValue("description") + " " + scenarioId);

    if (!builder.hasValue("roomPrice")) {
      builder.with(
          "roomPrice",
          TestData.numericSuffix(defaults.priceMinimum(), defaults.priceMaximumExclusive()));
    }

    return builder;
  }

  public RoomDataBuilder withRoomId(int roomId) {
    return with("roomid", roomId).with("roomName", Integer.toString(roomId));
  }

  public RoomDataBuilder withType(String type) {
    return with("type", type);
  }

  public RoomDataBuilder with(String propertyName, Object value) {
    payload.set(propertyName, valueToJson(value));
    return this;
  }

  public ObjectNode buildPayload() {
    return payload.deepCopy();
  }

  private String stringValue(String propertyName) {
    return JsonPayloads.stringValue(payload, propertyName);
  }

  private boolean hasValue(String propertyName) {
    return payload.has(propertyName) && !payload.get(propertyName).isNull();
  }

  private static JsonNode valueToJson(Object value) {
    return JSON.valueToTree(value);
  }
}

File: playwright-csharp/tests/Data/DataBuilders/RestfulBooker/RoomDataBuilder.cs
using System.Globalization;
using System.Text.Json.Nodes;
using TestTracks.Playwright.CSharp.Specs.Data;

namespace TestTracks.Playwright.CSharp.Specs.Data.DataBuilders.RestfulBooker;

public sealed record RoomDataDefaults
{
    public required JsonObject Payload { get; init; }

    public int PriceMinimum { get; init; } = 300;

    public int PriceMaximumExclusive { get; init; } = 900;

    public int RoomIdMinimum { get; init; } = 7000;

    public int RoomIdMaximumExclusive { get; init; } = 9999;
}

public sealed class RoomDataBuilder
{
    private readonly JsonObject _payload;

    public RoomDataBuilder(RoomDataDefaults defaults)
    {
        _payload = defaults.Payload.DeepClone().AsObject();
    }

    public static RoomDataBuilder ForScenario(string scenarioId, RoomDataDefaults defaults)
    {
        var roomNumber = TestData.NumericSuffix(defaults.RoomIdMinimum, defaults.RoomIdMaximumExclusive);

        var builder = new RoomDataBuilder(defaults);
        builder
            .WithRoomId(roomNumber)
            .With("description", $"{builder.StringValue("description")} {scenarioId}");

        if (!builder.HasValue("roomPrice"))
        {
            builder.With("roomPrice", Random.Shared.Next(defaults.PriceMinimum, defaults.PriceMaximumExclusive));
        }

        return builder;
    }

    public RoomDataBuilder WithRoomId(int roomId)
    {
        return With("roomid", roomId)
            .With("roomName", roomId.ToString(CultureInfo.InvariantCulture));
    }

    public RoomDataBuilder WithType(string type)
    {
        return With("type", type);
    }

    public RoomDataBuilder With(string propertyName, object? value)
    {
        _payload[propertyName] = JsonValue.Create(value);
        return this;
    }

    public JsonObject BuildPayload()
    {
        return _payload.DeepClone().AsObject();
    }

    private string StringValue(string propertyName)
    {
        return _payload[propertyName]?.GetValue<string>()
            ?? throw new InvalidOperationException($"Room payload is missing '{propertyName}'.");
    }

    private bool HasValue(string propertyName)
    {
        return _payload.ContainsKey(propertyName) && _payload[propertyName] is not null;
    }
}
